core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HttpHealthChecker.java

```diff
@@ -70,15 +88,40 @@ void start() {
     }
 
     private synchronized void check() {
-        lastCheckFuture = httpClient.head(path).aggregate().handle((res, cause) -> {
-            if (res != null && res.status().equals(HttpStatus.OK)) {
-                ctx.updateHealth(1);
+        final RequestHeaders headers;
+        final RequestHeadersBuilder builder =
+                RequestHeaders.builder(HttpMethod.HEAD, path)
+                              .add(HttpHeaderNames.AUTHORITY, ctx.endpoint().authority());
+        if (maxLongPollingSeconds > 0) {
+            headers = builder.add(HttpHeaderNames.IF_NONE_MATCH, wasHealthy ? "\"healthy\"" : "\"unhealthy\"")
+                             .add(HttpHeaderNames.PREFER, "wait=" + maxLongPollingSeconds)
+                             .build();
+        } else {
+            headers = builder.build();
+        }
+
+        lastResponse = httpClient.execute(headers);
+        lastResponse.aggregate().handle((res, cause) -> {
+            boolean isHealthy = false;
+            if (res != null) {
+                maxLongPollingSeconds = Math.max(0, res.headers().getLong(ARMERIA_LPHC, 0));
+                if (res.status().equals(HttpStatus.OK)) {
+                    isHealthy = true;
+                }
             } else {
-                ctx.updateHealth(0);
+                maxLongPollingSeconds = 0;
             }
+            ctx.updateHealth(isHealthy ? 1 : 0);
+            wasHealthy = isHealthy;
 
             try {
-                ctx.executor().schedule(this::check, ctx.nextDelayMillis(), TimeUnit.MILLISECONDS);
+                final ScheduledExecutorService executor = ctx.executor();
+                final long nextDelayMillis = ctx.nextDelayMillis();
+                if (res != null && maxLongPollingSeconds > 0) {
+                    executor.execute(this::check);
```

Comment:
I think we still need to schedule using `nextDelayMillis` to avoid aggressive health check.

Reply:
That's why I only do this when `res != null`. What other conditions could we add to make this safer?

Reply:
Ah yeah, that makes sense. Thanks!
By the way, we can move the `final long nextDelayMillis = ctx.nextDelayMillis();` to the bottom `else` block, or inline it.

Reply:
Updated the logic so that it enables long polling only on 2xx and 5xx responses, so that we do not hammer a server on 4xx, etc.

Reply:
That's a good idea. 👍